### PR TITLE
created check runner in bmptk

### PR DESCRIPTION
**Pull request by sqrtroot**
created check runner in bmptk so people can run our check suite
clang-tidy and cppcheck are supported at the moment

### CODE REVIEW

**Comment by Bob-Thomas** [comment]
+1
neat functionality so we can actually run checks without waiting on jenkins

**Comment by ManDeJan** [comment]
+1, good functionality. Code also looks clean, also worked as expected when I tested it. I approve for a merge.

**Review by ManDeJan — DISMISSED** [review]
Tested and approved for merge

File: bmptk++.py [hunk 1]
```diff
@@ -100,6 +101,93 @@ def generate(args):
...
+            self.description)
+
+
+def check_one(cc_db):
+    errors = []
+    with open(cc_db) as cc_db_file:
+        cc_json = json.load(cc_db_file)
+        module = cc_db.split("/")[-3]
+        for entry in cc_json:
+            tidy_command = 'clang-tidy-3.8 -header-filter=.* -p={0} -checks=* {1}'.format(
```
Comment (cjsmeele) [inline bmptk++.py]: Why depend on this specific clang-tidy version?
At least on my machine (Debian 8) `clang-tidy` is in my path, but this script won't run it as it is a different version.
(same issue on Arch where my clang-tidy is newer, `clang-tidy` itself is in path)

**Review by cjsmeele — CHANGES_REQUESTED** [review]
Looks good overall and adds a very useful feature!
There's only a small issue with selecting the right clang-tidy binary, see other comment.

File: bmptk++.py (hunk 1 above)
Comment (sqrtroot) [reply bmptk++.py]: This is how the official python script did it. I agree that it's not the correct way change is incoming :)

Comment (ManDeJan) [reply bmptk++.py]: clang-tidy-3.8 is a separate package then clang-tidy on my linux mint system so I thats why I did not encouter problems when testing

**Review by ManDeJan — DISMISSED** [review]
Requested changes have been made, ok to merge.

**Comment by cjsmeele** [comment]
I have an arduino module that has the following code in `main.cc`:
```c++
// Unused arguments
int f(int a, int b) { return 5; }                                                                                  

int main() {
    WDT->WDT_MR = WDT_MR_WDDIS;

    int x; // uninitialized variable passed to f
    int y;
    y = 5;
         y++; // (4 spaces + 1 tab char)
    f(x, y);

    // (...)
}
```

Running `bmptk++.py` with command `check -m GAS` gives the following output:

```
Command: check -m GAS
Checking module GAS
[Cppcheck]main.cc:14  Uninitialized variable
```

Can you confirm that it should only catch this one warning?
Should the other issues not be caught?

For reference, this is on Arch with clang-tidy 4.0 and cppcheck 1.79.

**Comment by sqrtroot** [comment]
@cjsmeele 
The current test runners (clang-tidy and cppcheck) do not check all style issues. Clang-format has to be implemented to generate these errors. The problem is that clang-format does not output the errors it has found in the styling only a formated version of the input file.

@R2D2-2017/leads Maybe we have to take a look at the implementation of clang-format to see if we can generate these kinds of errors ourself based on the tooling in llvm?

**Comment by cjsmeele** [comment]
@sqrtroot Agreed. I'll approve this PR with the understanding that we're going to extend this feature in the near future.

**Review by cjsmeele — DISMISSED** [review]

**Comment by sqrtroot** [comment]
After some more digging I found the reason the unused arguments are not found by clang-tidy. Apperently it doesn't go well with the generated compile_commands.json file and probably arm-none-eabi-g++ 
If i run ` clang-tidy -header-filter=.* -checks=* R2D2-2017/modules/TEST/src/main.cc` (main.cc contains the code mentioned above) it generates the following output:
```
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "R2D2-2017/modules/TEST/src/main.cc"
No compilation database found in R2D2-2017/modules/TEST/src or any parent directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
2 warnings and 2 errors generated.
Error while processing R2D2-2017/modules/TEST/src/main.cc.
R2D2-2017/modules/TEST/src/main.cc:2:11: warning: parameter 'a' is unused [misc-unused-parameters]
int f(int a, int b) { return 5; }                                                                                  
          ^
           /*a*/
R2D2-2017/modules/TEST/src/main.cc:2:18: warning: parameter 'b' is unused [misc-unused-parameters]
int f(int a, int b) { return 5; }                                                                                  
                 ^
                  /*b*/
R2D2-2017/modules/TEST/src/main.cc:5:5: error: use of undeclared identifier 'WDT' [clang-diagnostic-error]
    WDT->WDT_MR = WDT_MR_WDDIS;
    ^
R2D2-2017/modules/TEST/src/main.cc:5:19: error: use of undeclared identifier 'WDT_MR_WDDIS' [clang-diagnostic-error]
    WDT->WDT_MR = WDT_MR_WDDIS;

```
so without the compilation database it does find the unused parameters but it errors on undeclared identifiers

**Comment by sqrtroot** [comment]
It took some time and is a bit hacky but luckily it finally works for arduinos.
Please review @cjsmeele @ManDeJan

**Review by ManDeJan — APPROVED** [review]
If it's hacky and you know it clap your hands *clap* *clap*
If you aprove for merging and you know it clap your hands  *clap* *clap*'

I don't see how else we could get this working in time and it works, and nobody has to see or touch it, so I approve.
Nice work :) 👍

**Pull request merged**